lib/devise_security_extension/hooks/session_non_transferable.rb

```diff
@@ -0,0 +1,33 @@
+# This hook requires session_limitable 
+
+# After each sign in, native Devise updates current_sign_in_ip and this hook
+# updates unique_session_client.
+# This is only triggered when the user is explicitly set (with set_user)
+# and on authentication. Retrieving the user from session (:fetch) does
+# not trigger it.
+Warden::Manager.after_set_user :except => :fetch do |record, warden, options|
+  if record.respond_to?(:update_current_user_agent!) && warden.authenticated?(options[:scope])
```

Comment:
`respond_to?`, `kind_of?`, and related methods are generally anti-pattern in Ruby, because you can use duck typing and polymorphism to determine how an object should respond.